src/compute-client/src/plan/interpret/physically_monotonic.rs

```diff
@@ -7,5 +7,245 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-//! Placeholder module for an [crate::plan::interpret::Interpreter] that infers
-//! physical monotonicity.
+//! Implementation of [crate::plan::interpret::Interpreter] for inference
+//! of physical monotonicity in single-time dataflows.
+
+use std::collections::BTreeSet;
+use std::marker::PhantomData;
+
+use differential_dataflow::lattice::Lattice;
+use mz_expr::{EvalError, Id, MapFilterProject, MirScalarExpr, TableFunc};
+use mz_repr::{Diff, GlobalId, Row};
+use timely::PartialOrder;
+
+use crate::plan::interpret::{BoundedLattice, Context, Interpreter};
+use crate::plan::join::JoinPlan;
+use crate::plan::reduce::{KeyValPlan, ReducePlan};
+use crate::plan::threshold::ThresholdPlan;
+use crate::plan::top_k::TopKPlan;
+use crate::plan::{AvailableCollections, GetPlan};
+
+/// Represents a boolean physical monotonicity property, where the bottom value
+/// is true (i.e., physically monotonic) and the top value is false (i.e. not
+/// physically monotonic).
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct PhysicallyMonotonic(pub bool);
+
+impl BoundedLattice for PhysicallyMonotonic {
+    fn top() -> Self {
+        PhysicallyMonotonic(false)
+    }
+
+    fn bottom() -> Self {
+        PhysicallyMonotonic(true)
+    }
+}
+
+impl Lattice for PhysicallyMonotonic {
+    fn join(&self, other: &Self) -> Self {
+        PhysicallyMonotonic(self.0 && other.0)
+    }
+
+    fn meet(&self, other: &Self) -> Self {
+        PhysicallyMonotonic(self.0 || other.0)
+    }
+}
+
+impl PartialOrder for PhysicallyMonotonic {
+    fn less_equal(&self, other: &Self) -> bool {
+        // We wish to implement the following truth table
```

Comment:
Can you explain the rationale to go with the `std::cmp::Reverse<bool>` instead of just the regular `bool` ordering here?

At first I thought it must have something to do with the way the lattice is used in `LetRec` blocks, but in those blocks you just say "I give up" based on the `is_rec` context below.

Reply:
Good observation, changed here to just use `Reverse::<bool>` comparison directly. I also added a comment to remark that the implementation is for consistency with the definitions for `top()` and `bottom()`. AFAICT we do not really need to make use of `less_equal` at present.